### ISSUE TITLE
DbConnection

### ISSUE THREAD

**Issue by Wayne-Mather** [roadrunnerlenny/etlbox#1]
Hi,

I was just starting to look at your code as I am about to build a DW and decided on .NET core for performance of in-memory operations. Was looking at your code and some interesting stuff in there.

I have not completely gone through it all but was just looking at your DbConnection class and specifically Open(). It looks like you expect your end users to know what they are doing and do things the right way.

As an example from your demo, what if the developer does this:

```
ControlFlow.CurrentDbConnection.Open();
// maybe more code, the issue is calling Open() more than once
ControlFlow.CurrentDbConnection.Open();
```
Same issue with your ExecuteXXX() statements. There is no using block to dispose of the Command object.

```
public IDataReader ExecuteReader(string commandText, IEnumerable<QueryParameter> parameterList = null) {
  Command cmd = CreateCommand(commandText, parameterList);
  return cmd.ExecuteReader();
}
```

Just wondering if you should have more defensive code to stop potential memory leaks? 

Hopefully my message does not come across as something negative  :)

**Comment by roadrunnerlenny** [comment]
Hi,

thank you for your constructive feedback! I am happy that you were able to find some interesting parts in my code. 

I just looked at the `DBConnection` class again, and it looks like this:

```C#
public void Open() {
        ....
        if (!IsConnectionOpen) {
              ...
        }
    }
}
```

So actually the developer could call Open() as often as he wants to, if a connection is open there won't be a new connection reopened. 

But you are right, the ExecuteXXX - Statements do not explicitly open/close a connection nor do they have a using statement in there. This is actually on purpose - the DBConnection (which is marked as abstract, so the inherited `SqlConnectionManager`) should not be called directly or used at all.

Let me explain the idea behind this:  If you look at the the class ```DBTask``` which is using the the ```DBConnectionManager```, you will see that all the ExecuteXXX - methods include the corresponding using and Open - method. E.g.:

```C#
public int ExecuteNonQuery() {
     using (var conn = DbConnectionManager.Clone()) {
          conn.Open();
          ...
       } ...
}
````
The reason behind this is that the ConnectionManager classes itself were never intended to be used directly. Their purpose is to be only handed over to the Tasks classes, e.g. `SqlTask`. The SqlTask will then use the connection Manager to open the connection, execute a ADO.NET command and then close the connection. Also, it will do this operations never on the connection manager itself, it will always use a clone of it. The clone is necessary because the whole ConnectionManager is just a wrapper for the ADO.NET classes. And ADO.NET has it own way of connection pooling which we want to use  - [see there microsoft documentation](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql-server-connection-pooling) for more details about connection pooling).  If SqlTask would be called "in parallel" (e.g. in a `Parallel.ForEach` - Loop), the clone guarantees that we always get a new Connection manager instance - but the ADO.NET connection pooling would take care of the real connection management. 

So, basically there could be 2 things changed or added to current code
* The Visibility of the Open/Close Methods in the `DBConnectionManager` class are changed to `internal`
* There are one or more tests added that proof that the connection pooling of ADO.NET is working and there are no memory leaks

What do you think? Is that what you were pointing out or do I miss something important here? Or did I understand your question wrong?

**Comment by Wayne-Mather** [comment]
Hi,

I think you understood what I was saying and there is a potential I am wrong and not understanding your code correctly.

In DbConnectionManager the code looks like this:

```
public void Open() {
  DbConnection = new Connection();
  if(!IsConnectionOpen) {
   // rest of code

```

Would this not create a new Connection() object each time? From your example code that object would be a new SqlConnection object and this object implements IDisposable. 

So correct me if I am wrong but is the code not written the way your intend? 

1. User calls Open()
2. This first call creates a SqlConnection() object
2. It then checks to ensure that it's not open and attempts to open a connection
3. User calls Open() again
4. This second call creates a new SqlConnection() object
5. It then checks to ensure that it's not open, which it is not because its a new one...

Hopefully these 5 steps help you understand what I think I am seeing with your code. I don't mind if there is something I am missing. 

I have just cloned the code to my local workstation and will write some unit tests to prove or disprove my theory, but that is what I see just by reading the code.

**Comment by roadrunnerlenny** [comment]
Hi,
thank you for the clarification . Now I think I really got your point.
You basically pointed out that the check `IsConnectionOpen` is actually a check on the `DBConnection` object: `DbConnection?.State == ConnectionState.Open;` So doing this check right after creating a new instance is more or less useless - a new `DBConnection` object (especially without any ConnectionString parameter set) will never be in the open state. So you are right, the if-clause could (and should) be removed and doesn't really do what my intention was. 

I just created a new Branch "Issue1" where I commented out the if-clause. 

So very likely if a developer  would call the Open - method again and again, a DB connection would open up every time. This is actually more or less the same behaviour you get when you use ADO.NET directly - if you use the `DBConnection` from `System.Data`, whenever a developer opens a connection he/she needs to close it in the code. That's why the `DBConnectionManager` is implementing the `IDisposable` interface, so  a developer could do something like this:

```C#
using (ControlFlow.CurrentDbConnection)
{
                ControlFlow.CurrentDbConnection.Open();
} 
```

but actually this is kind of "crappy" - anyone who would call Open() on an object store in a static property twice would except the second call to have no affect. So basically you found a bug here, because somehow the Open - method behaves "weird".

So I would be very happy if you would submit a new test as you described above, to test what are consequences of calling Open() without closing first. 

And if I get you right you are suggesting to improve this behaviour - calling the open method a second time should test first if the connection was closed (or disposed and set to null). If the connection wasn't closed/disposed, the class should do this for you before opening another one. Do you have any ideas how to implement this? Actually this would be a better behaviour than the current one, but right now I am not quite sure how to code this.



Regarding your last part,

>  I believe a clean external API that conveys meanings such as that no constant manual reading is better

I totally agree - the code and the API should always be more or less self-explaining. The question which needs to be answered is if there is an advantage of use the `DBConnectionManager` (or the inherited classes like `SqlConnectionManager`). The idea behind the library was that the Tasks should do all the work - so the `SqlTask` would need a ConnectionManager. That's why I wanted to set the methods in the `DBConnectionManager` to internal. But I am only one person, and so I have limited imagination what other people expect from the library. 
So what do you think: Would it be helpful to leave the `DBConnectionManager` public? So a developer could use the Open/Close and ExecuteXXX methods in it? Or would it be better to hide them (make them internal) and force the developer to use the corresponding tasks like `SqlTask`?

Again, I really appreciate that you take the time to discuss with me and review the code.

**Comment by Wayne-Mather** [comment]
Hi,

These sorts of problems are just experienced once you start writing code others use :)

The same problem can be thought of about your connection string. If I change the connection string on the manager what is the expectation? 


```
// simple fix, make the connection string private so only injected at construction
public ConnectionString ConnectionString { get; private set; }

```

For your open method I would write something like below - which I have also made it threadsafe-ish...

```
private object _lock = new object();

public void Open() {
    lock (_lock)
    {
        if (DbConnection == null)
        {
            DbConnection = new Connection();
        }

        if (!IsConnectionOpen)
        {
            DbConnection.ConnectionString = ConnectionString.Value;
            bool successfullyConnected = false;
            Exception lastException = null;
            for (int i = 1; i <= MaxLoginAttempts; i++)
            {
                try
                {
                    DbConnection.Open();
                    successfullyConnected = true;
                }
                catch (Exception e)
                {
                    successfullyConnected = false;
                    lastException = e;
                    Task.Delay(500 * i).Wait();
                }

                if (successfullyConnected) break;
            }

            if (!successfullyConnected)
                throw lastException ?? new Exception("Could not connect to database!");
        }
    }
}
```

I have looked more at your code from a high-level and I see you are essentially trying to emulate SSIS without a canvas. 

I have a different vision of what I want so will not use ETLBox for my needs. The reason is I want an architecture that is Job based with threading. I also want to be able to just drop DLL's into a folder to add/update tasks as required. I want my framework to essentially allow the user to do anything in the jobs but they have the ability to execute parallel tasks.

I did create software that automated the generation of Stored Procedures and Tables from meta data. Like you I found that depending on the staging requirements, looping and updating data in an ACID compliant data store not only decreases performance, but cannot be easily tested using modern development techniques.

However, keep going on your project because in dotnet core 3 you have WPF and you could create graphical tools to automate the code generation (like EF Designers). 

I do believe however, that at some point Microsoft may just rewrite SSIS on top of dotnet core so that the packages can run directly on Linux servers, or, come out with something better.

**Comment by roadrunnerlenny** [comment]
Hi,
thank you for your feedback and good advice. I will make a test for the open-connection method as you described it and very likely do some changes to the open-method and the `DBConnectionManager` class as you suggested. Also your comment about the `ConnectionString` is very helpful, I  will have a look into this. Thank you for the good input!

Regarding your general comment about the library - it is very interesting to hear what other developers are looking for when it comes to an ETL library. I've seen different requirements and implementations of ETL frameworks so far, and SSIS is a toolset which is very "complete". That's why I used it as a template for my library, hoping that when other developers will start using my library they will recognize stuff and it will perhaps make it easier for them. But after all, ETLBox is based on two basic technologies from Microsoft - ADO.NET and the TPL.Dataflow library. Both are essentially the two core frameworks that do all the "hard work". I just tried to put them together in a way that they can be easier used programmatically when you are trying to create ETL jobs with C#.

I wish you good luck finding the right toolset for your needs. Again, I appreciate that you took the time to review my code and to give me some very helpful feedback.

**Comment by anlennar** [comment]
I did some smaller changes to the DBConnectionManager class. The fix is merged into the release branch and will be part of the next version.

I also added some tests, to verify that the DBConnectionManager is working as expected.

Here the case where one connection is opened and then close - the connection pooling will normally keep the connection alive, in case it could be needed again.

```C#
public void TestOpeningCloseConnection() {
    SqlConnectionManager con = new SqlConnectionManager(new ConnectionString(ConnectionStringParameter));       
   AssertOpenConnectionCount(0);
   con.Open();
   AssertOpenConnectionCount(1);
   con.Close(); //won't close any connection - ado.net will keep the connection open in it's pool in case it's needed again
  AssertOpenConnectionCount(1);
  SqlConnection.ClearAllPools();
  AssertOpenConnectionCount(0);
}
```

When trying to open a connection twice, the connection manager will just tell ADO.NET to reopen a connection with the exact same parameter. ADO.NET will not open a new connection cause it will recognize that the parameter are the same.

```C#
public void TestOpeningConnectionTwice() {
            SqlConnectionManager con = new SqlConnectionManager(new ConnectionString(ConnectionStringParameter));
            AssertOpenConnectionCount(0);
            con.Open();
            con.Open();
            AssertOpenConnectionCount(1);
            con.Close();
            AssertOpenConnectionCount(1);
            SqlConnection.ClearAllPools();
            AssertOpenConnectionCount(0);
}
```

I also added a test which execute Sql Statements on the database in parallel (using TPL). Each thread will get it's own connection from the ADO.NET pool.

The idea of making the Open-method and the ConnectionString property `internal` did not work out - the interface implementation prevented this.

**Comment by roadrunnerlenny** [comment]
I did some smaller changes to the DBConnectionManager class. The fix is merged into the release branch and will be part of the next version.

I also added some tests, to verify that the DBConnectionManager is working as expected.

Here the case where one connection is opened and then close - the connection pooling will normally keep the connection alive, in case it could be needed again.

```C#
public void TestOpeningCloseConnection() {
    SqlConnectionManager con = new SqlConnectionManager(new ConnectionString(ConnectionStringParameter));       
   AssertOpenConnectionCount(0);
   con.Open();
   AssertOpenConnectionCount(1);
   con.Close(); //won't close any connection - ado.net will keep the connection open in it's pool in case it's needed again
  AssertOpenConnectionCount(1);
  SqlConnection.ClearAllPools();
  AssertOpenConnectionCount(0);
}
```

When trying to open a connection twice, the connection manager will just tell ADO.NET to reopen a connection with the exact same parameter. ADO.NET will not open a new connection cause it will recognize that the parameter are the same.

```C#
public void TestOpeningConnectionTwice() {
            SqlConnectionManager con = new SqlConnectionManager(new ConnectionString(ConnectionStringParameter));
            AssertOpenConnectionCount(0);
            con.Open();
            con.Open();
            AssertOpenConnectionCount(1);
            con.Close();
            AssertOpenConnectionCount(1);
            SqlConnection.ClearAllPools();
            AssertOpenConnectionCount(0);
}
```

I also added a test which execute Sql Statements on the database in parallel (using TPL). Each thread will get it's own connection from the ADO.NET pool.

The idea of making the Open-method and the ConnectionString property `internal` did not work out - the interface implementation prevented this.

**Comment by roadrunnerlenny** [comment]
Hi,
you are right, the dispose of the command is currently not happening. I did a quick research on google, it looks like that the dispose function of the command does not really do much. But still it is a `IDisposable` object and it should be either wrapped in a using block or the dispose method should be called. So I will definitively add this to the ExecuteXXX functions, thanks for the hint!
